aten/src/ATen/native/cuda/Indexing.cu

```diff
@@ -123,6 +126,55 @@ __global__ void indexing_backward_kernel(
 
 namespace at { namespace native {
 
+namespace {
+
+class ReduceMultiply {
+public:
+  template <typename scalar_t>
+  constexpr C10_DEVICE void operator() (scalar_t * self_data, const scalar_t * src_data) const {
+    gpuAtomicMul(self_data, *src_data);
+  }
+};
+static ReduceMultiply reduce_multiply;
+
+class ReduceAdd {
+public:
+  template <typename scalar_t>
+  constexpr C10_DEVICE void operator() (scalar_t * self_data, const scalar_t * src_data) const {
+    gpuAtomicAddNoReturn(self_data, *src_data);
```

Comment:
You might want to start using `fastAtomicAdd` before people start complaining about bad fp16 performance. It is a slightly difference API though as it requires sending base address of a tensor and an index to update.